src/transforms/babel/env.js

```diff
@@ -0,0 +1,65 @@
+const presetEnv = require('@babel/preset-env');
+const getTargetEngines = require('../../utils/getTargetEngines');
+
+/**
+ * Generates a @babel/preset-env config for an asset.
+ * This is done by finding the source module's target engines, and the app's
+ * target engines, and doing a diff to include only the necessary plugins.
+ */
+async function getEnvConfig(asset, isSourceModule) {
+  // Load the target engines for the app and generate a @babel/preset-env config
+  let targetEngines = await getTargetEngines(asset, true);
+  let targetEnv = await getEnvPlugins(targetEngines, true);
+  if (!targetEnv) {
+    return null;
+  }
+
+  // If this is the app module, the source and target will be the same, so just compile everything.
+  // Otherwise, load the source engines and generate a babel-present-env config.
+  if (!isSourceModule) {
+    let sourceEngines = await getTargetEngines(asset, false);
+    let sourceEnv = (await getEnvPlugins(sourceEngines, false)) || targetEnv;
+
+    // Do a diff of the returned plugins. We only need to process the remaining plugins to get to the app target.
+    let sourcePlugins = new Set(sourceEnv.map(p => p[0]));
+    targetEnv = targetEnv.filter(plugin => {
+      return !sourcePlugins.has(plugin[0]);
+    });
+  }
+
+  return {
+    internal: true,
+    babelVersion: 7,
+    config: {
+      plugins: targetEnv,
+    }
+  };
+}
+
+const envCache = new Map();
+
+async function getEnvPlugins(targets, useBuiltIns = false) {
+  if (!targets) {
+    return null;
+  }
+
+  let key = JSON.stringify(targets);
+  if (envCache.has(key)) {
+    return envCache.get(key);
+  }
+
+  let plugins = presetEnv.default(
+    {assertVersion: () => true},
```

Comment:
Is there a helper function to run a preset function that I can use instead? I just need the list of plugins it produces.

Reply:
Could you point me to where you are processing these plugins? I'm having trouble finding it.

Part of my issue here is that the exact return value of a preset is not a public API, because a preset can also have other presets, and what we'd actually need is an API to fully flatten the configuration into a list of plugins, which is pretty much what `loadOptions` already provides. We wouldn't want to expose `makeAPI` because it exposes internal implementation details of Babel itself, and it's not obvious what anyone would actually be able to do with the result anyway, other than pass it right back to Babel again unprocessed.